pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,449 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package msgconnpool
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	v2 "mosn.io/mosn/pkg/config/v2"
+	"mosn.io/mosn/pkg/upstream/cluster"
+
+	"mosn.io/api"
+	"mosn.io/mosn/pkg/log"
+	"mosn.io/mosn/pkg/types"
+	"mosn.io/pkg/buffer"
+	"mosn.io/pkg/utils"
+)
+
+// for reconnect
+const (
+	notConnecting = iota
+	connecting
+)
+
+// types.ConnectionPool
+type connpool struct {
+	idleClients []*activeClient
+
+	host      types.Host
+	clientMux sync.Mutex
+
+	autoReconnectWhenClose bool
+	heartBeatCreator       func() KeepAlive
+	reconnTryTimes         int
+	readFilters            []api.ReadFilter
+
+	destroyed uint64
+}
+
+func (p *connpool) Host() types.Host {
+	return p.host
+}
+
+// GetActiveClient get a avail client
+func (p *connpool) GetActiveClient(ctx context.Context) (*activeClient, types.PoolFailureReason) {
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	n := len(p.idleClients)
+
+	// no available client
```

Comment:
The location of the code comment might be wrong

Reply:
fixed